lib/cfndsl/specification.rb

```diff
@@ -0,0 +1,153 @@
+          elsif property_info['ItemType']
+            # Tag is a reused type, but not quite primitive
+            # and not all resources use the general form
+            property_type = if property_info['ItemType'] == 'Tag'
+                              'Tag'
+                            else
+                              Array(resource_name.split('::').join + property_info['ItemType'])
+                            end
+          elsif property_info['Type']
+            # Special types (defined below) are joined with their parent
+            # resource name for uniqueness and connection
+            property_type = resource_name.split('::').join + property_info['Type']
+          else
+            warn "could not extract type from #{resource_name}"
+          end
+          extracted[property_name] = property_type
+          extracted
+        end
+        resources[resource_name] = { 'Properties' => properties }
+        resources
+      end
+    end
+
+    def self.extract_types(spec)
+      primitive_types = {
+        'String'    => 'String',
+        'Boolean'   => 'Boolean',
+        'Json'      => 'Json',
+        'Integer'   => 'Integer',
+        'Number'    => 'Number',
+        'Double'    => 'Double',
+        'Timestamp' => 'Timestamp',
+        'Long'      => 'Long'
+      }
+      spec.each_with_object(primitive_types) do |(property_name, property_info), types|
+        # In order to name things uniquely and allow for connections
+        # we extract the resource name from the property
+        # AWS::IAM::User.Policy becomes AWSIAMUserPolicy
+        root_resource = property_name.match(/(.*)\./)
+        root_resource_name = root_resource ? root_resource[1].gsub(/::/, '') : property_name
+        property_name = property_name.gsub(/::|\./, '')
+        next unless property_info['Properties']
+        properties = property_info['Properties'].each_with_object({}) do |(nested_prop_name, nested_prop_info), extracted|
+          if nested_prop_info['Type'] == 'Map' || nested_prop_info['Type'] == 'Json'
+            # The Map type and the incorrectly labelled Json type
+            nested_prop_type = 'Json'
+          elsif nested_prop_info['PrimitiveType']
+            nested_prop_type = nested_prop_info['PrimitiveType']
+          elsif nested_prop_info['PrimitiveItemType']
+            nested_prop_type = Array(nested_prop_info['PrimitiveItemType'])
+          elsif nested_prop_info['ItemType']
+            nested_prop_type = root_resource_name + nested_prop_info['ItemType']
+          elsif nested_prop_info['Type']
+            nested_prop_type = root_resource_name + nested_prop_info['Type']
+          else
+            warn "could not extract type from #{property_name}"
+          end
+          extracted[nested_prop_name] = nested_prop_type
+          extracted
+        end
+        types[property_name] = properties
+        types
+      end
+    end
+
+    # Missing/malformed resources from the resource specification
```

Comment:
@gergnz @kornypoet could I suggest moving this to a more common area like we did for fixing Parameter and Output?